### PR TITLE
Feature/14/14 verify JWS (with ADR)

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

Access to the Registration Service API is authenticated with JWS, using scheme and libraries provided in EDC (those libraries are used in EDC for IDS authentication).

See ADR for more details.

## Why it does that

Authentication ensures that only actors who can assert their control of a did:web identifier can register that identifier as a dataspace participant. 

## Linked Issue(s)

https://github.com/agera-edc/MinimumViableDataspace/issues/14

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: docs/developer/decision-records/2022-07-01-service-authentication/README.md
Comment (Izzzu) [inline docs/developer/decision-records/2022-07-01-service-authentication/README.md]:
```suggestion
The choice of authentication scheme and key type is driven by the schemes [currently supported in EDC](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/89ffb983dbdebce02e1447d2f2f5f65843f46041/docs/developer/decision-records/2022-06-19-json-web-token/README.md), and can be extended as needed in the future.
```

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java
Comment (Izzzu) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java]: An idea to include failure messages in the exception message, can be useful to see the specific failure message when the exception is thrown:
```
String failureMessages = String.join(" | ", verificationResult.getFailureMessages());
throw authenticationFailure(failureMessages);
```

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java
Comment (Izzzu) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java:109]: It's not straightforward to understand what's happening in this method step by step. We could improve the readability a bit by extracting some logical steps to methods. (Then if someone reading the code doesn't need to understand the details to have the overview of the logic)
Example:
```java
    @Override
    public void filter(ContainerRequestContext requestContext) {
        var headers = requestContext.getHeaders();
        Objects.requireNonNull(headers, "headers");

        String credential = getCredential(headers);
        SignedJWT jwt = getSignedJWT(credential);
        String issuer = getIssuer(jwt);
        validate(jwt, issuer);

        headers.putSingle(CALLER_DID_HEADER, issuer);
    }

    private String getCredential(MultivaluedMap<String, String> headers) {
        var authHeader = headers.getFirst("Authorization");
        if (authHeader == null) {
            throw authenticationFailure("Missing Authorization header");
        }
        var separatedAuthHeader = authHeader.split(" ");

        if (separatedAuthHeader.length != 2 || !"Bearer".equals(separatedAuthHeader[0])) {
            throw authenticationFailure("Authorization header value is not a valid Bearer token");
        }

        return separatedAuthHeader[1];
    }

    @NotNull
    private SignedJWT getSignedJWT(String credential) {
        SignedJWT jwt;
        try {
            jwt = SignedJWT.parse(credential);
        } catch (ParseException e) {
            throw authenticationFailure("Invalid JWT (parse error)");
        }
        return jwt;
    }

    private String getIssuer(SignedJWT jwt) {
        String issuer;
        try {
            issuer = jwt.getJWTClaimsSet().getIssuer();
        } catch (ParseException e) {
            throw authenticationFailure("Invalid JWT (parse error)");
        }
        return issuer;
    }

    private void validate(SignedJWT jwt, String issuer) {
        var publicKey = didPublicKeyResolver.resolvePublicKey(issuer);

        var verificationResult = VerifiableCredentialFactory.verify(jwt, publicKey.getContent(), audience);
        if (verificationResult.failed()) {
            verificationResult.getFailureMessages().forEach(message -> monitor.debug(() -> "Invalid JWT (verification error). " + message));
            throw authenticationFailure("Invalid JWT (verification error)");
        }

        if (publicKey.failed()) {
            publicKey.getFailureMessages().forEach(message -> monitor.debug(() -> "Failed obtaining public key for DID: " + issuer + ". " + message));
            throw authenticationFailure("Failed obtaining public key for DID");
        }

        monitor.debug("Valid JWT");
    }
```

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java
Comment (algattik) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java]: Thanks, implemented along similar lines

File: docs/developer/decision-records/2022-07-01-service-authentication/README.md
Comment (ouphi) [inline docs/developer/decision-records/2022-07-01-service-authentication/README.md]: I have a few questions:
What is an audience? How many different audience would we have inside one dataspace?
Why is this called audience if this is supposed to be a registration-service endpoint?

Why a caller would need to specify a registration service endpoint in a token when calling the registration service itself already?

**Review by Izzzu — APPROVED** [review]

**Review by cpeeyush — APPROVED** [review]
Approving as I was a reviewer before. But will wait for one more approval as I am taking care of this PR now. Thanks

File: docs/developer/decision-records/2022-07-01-service-authentication/README.md
Comment (cpeeyush) [reply docs/developer/decision-records/2022-07-01-service-authentication/README.md]: - audience is a claim as `aud` to identify the recipient for which a JWT is intended for. https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3 So I this case when a participant will interact with registration service then registration service is the audience.
- This claim is used here to make sure that JWT is intended for registration service which is verifying the signature and not for someone else.

Comment (ouphi) [reply docs/developer/decision-records/2022-07-01-service-authentication/README.md]: I see thanks!

Comment (ouphi) [inline docs/developer/decision-records/2022-07-01-service-authentication/README.md]:
```suggestion
| Audience        | The Registration Service API base endpoint (example: `http://dataspace.example.com:8182/authority`). This allows the server to verify the intended [audience](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3). |
```

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java
Comment (ouphi) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/auth/DidJwtAuthenticationFilter.java:58]: The 3 methods `parseJsonWebToken`, `getIssuerClaim` and `verifyTokenSignature` could be reused to verify credentials in the identity hub CredentialsVerifier.
No action needed right now, but just putting the information here in case we want to discuss how to share this code.

**Review by ouphi — APPROVED** [review]

**Pull request merged**
